src/rhsmlib/dbus/base_object.py

```diff
@@ -0,0 +1,208 @@
+
+
+class ReadWriteProperties(BaseProperties):
+    """A read-write BaseProperties.
+
+    Use this if you want to be able to set()/Set() DBus.Properties
+    on a service."""
+
+    def __setitem__(self, key, value):
+        if self.access_mask(key) != 'write':
+            raise exceptions.AccessDenied(key, self.interface_name)
+        self[key] = value
+
+    # FIXME: track read/write per property
+    def access_mask(self, prop_key):
+        return 'write'
+
+
+class BaseObject(dbus.service.Object):
+    # Name of the DBus interface provided by this object
+    interface_name = constants.INTERFACE_BASE
+    default_dbus_path = constants.ROOT_DBUS_PATH
+
+    def __init__(self, conn=None, object_path=None, bus_name=None):
+        super(BaseObject, self).__init__(conn=conn, object_path=object_path, bus_name=bus_name)
+        self.object_path = self.default_dbus_path
+
+        # TODO: Support multiple interfaces
+        self.properties = {self.interface_name: self._create_properties(self.interface_name)}
+
+    def _create_properties(self, interface_name):
+        return BaseProperties.create_instance(interface_name, {}, self.PropertiesChanged)
+
+    def _check_interface(self, interface_name):
+        if interface_name != self.interface_name:
+            raise exceptions.UnknownInterface(interface_name)
+
+    @dbus.service.signal(dbus.PROPERTIES_IFACE, signature='sa{sv}as')
+    def PropertiesChanged(self, interface_name, changed_properties, invalidated_properties):
+        # TODO: Do something here
+        log.debug("Properties Changed emitted.")
+
+    @util.dbus_service_method(
+        dbus.PROPERTIES_IFACE,
+        in_signature='s',
+        out_signature='a{sv}')
+    @util.dbus_handle_exceptions
+    def GetAll(self, interface_name, sender=None):
+        self._check_interface(interface_name)
+        return dbus.Dictionary(self.properties[interface_name].get_all(), signature='sv')
+
+    @util.dbus_service_method(
+        dbus.PROPERTIES_IFACE,
+        in_signature='ss',
+        out_signature='v')
+    @util.dbus_handle_exceptions
+    def Get(self, interface_name, property_name, sender=None):
+        self._check_interface(interface_name)
+        return self.properties[interface_name][property_name]
+
+    @util.dbus_service_method(
+        dbus.PROPERTIES_IFACE,
+        in_signature='ssv')
+    @util.dbus_handle_exceptions
+    def Set(self, interface_name, property_name, new_value, sender=None):
+        """Set a DBus property on this object.
+
+        This is the base service class, and defaults to DBus properties
+        being read-only. Attempts to Set a property will raise a
+        DBusException of type org.freedesktop.DBus.Error.AccessDenied.
+
+        Subclasses that need settable properties should override this."""
+        self._check_interface(interface_name)
+        self.properties[interface_name][property_name] = new_value
```

Comment:
Some of this stuff could probably just go in `src/rhsmlib/dbus/__init__.py` unless you see a reason not to? I find `base_object.BaseObject` a bit odd; but not strongly against it.